File: src/domainwizard/fileextent.py
"""Reads the extent and CRS of a raster or vector file via GDAL/OGR directly.

Replaces the QGIS-plugin's "Set to Active Layer Extent" button, which relied
on QGIS's layer list (iface.activeLayer()). There's no such concept outside
QGIS, so instead the user picks a file and we read its extent with GDAL/OGR.
"""

from typing import Tuple

from osgeo import gdal, ogr, osr

from gis4wrf.core import BoundingBox2D


def read_extent_and_srs(path: str) -> Tuple[BoundingBox2D, osr.SpatialReference]:
    ds = gdal.OpenEx(path, gdal.OF_RASTER | gdal.OF_VECTOR)
    if ds is None:
        raise ValueError(f'Could not open {path} as a raster or vector file')

    if ds.RasterCount > 0:
        gt = ds.GetGeoTransform()
        width, height = ds.RasterXSize, ds.RasterYSize
        x0, y0 = gt[0], gt[3]
        x1 = gt[0] + width * gt[1] + height * gt[2]
        y1 = gt[3] + width * gt[4] + height * gt[5]
        bbox = BoundingBox2D(minx=min(x0, x1), maxx=max(x0, x1), miny=min(y0, y1), maxy=max(y0, y1))
        srs = osr.SpatialReference()
        srs.ImportFromWkt(ds.GetProjection())
        return bbox, srs

    layer = ds.GetLayer(0)
    if layer is None:
        raise ValueError(f'{path} has no raster bands and no vector layers')
    xmin, xmax, ymin, ymax = layer.GetExtent()
    bbox = BoundingBox2D(minx=xmin, maxx=xmax, miny=ymin, maxy=ymax)
    srs = layer.GetSpatialRef()
    if srs is None:
        raise ValueError(f'{path} has no spatial reference defined')
    return bbox, srs

File: cpp/src/file_extent.cpp
#include "wrftools/file_extent.hpp"
#include "wrftools/error.hpp"

#include <gdal_priv.h>
#include <ogrsf_frmts.h>

#include <algorithm>

namespace wrftools {

FileExtent readFileExtent(const std::filesystem::path& path) {
    GDALAllRegister();
    std::unique_ptr<GDALDataset, void (*)(GDALDataset*)> dataset(
        static_cast<GDALDataset*>(GDALOpenEx(path.string().c_str(), GDAL_OF_RASTER | GDAL_OF_VECTOR, nullptr, nullptr, nullptr)),
        [](GDALDataset* d) { if (d) GDALClose(d); });
    if (!dataset) throw UserError("Could not open " + path.string() + " as a raster or vector file");

    if (dataset->GetRasterCount() > 0) {
        double gt[6];
        if (dataset->GetGeoTransform(gt) != CE_None) throw UserError(path.string() + " has no geotransform.");
        const double width = dataset->GetRasterXSize(), height = dataset->GetRasterYSize();
        const double x0 = gt[0], y0 = gt[3];
        const double x1 = gt[0] + width * gt[1] + height * gt[2];
        const double y1 = gt[3] + width * gt[4] + height * gt[5];
        Bounds2D bounds{std::min(x0, x1), std::min(y0, y1), std::max(x0, x1), std::max(y0, y1)};
        const auto* wkt = dataset->GetProjectionRef();
        if (!wkt || !*wkt) throw UserError(path.string() + " has no spatial reference defined.");
        return {bounds, Crs::fromWkt(wkt)};
    }

    auto* layer = dataset->GetLayer(0);
    if (!layer) throw UserError(path.string() + " has no raster bands and no vector layers");
    OGREnvelope envelope;
    if (layer->GetExtent(&envelope) != OGRERR_NONE) throw UserError("Could not read the extent of " + path.string());
    auto* srs = layer->GetSpatialRef();
    if (!srs) throw UserError(path.string() + " has no spatial reference defined.");
    char* wkt = nullptr;
    srs->exportToWkt(&wkt);
    std::string wktString = wkt ? wkt : "";
    CPLFree(wkt);
    return {{envelope.MinX, envelope.MinY, envelope.MaxX, envelope.MaxY}, Crs::fromWkt(wktString)};
}

}  // namespace wrftools
